Scalar.Common/Maintenance/ConfigStep.cs

```diff
@@ -208,25 +209,86 @@ public bool TrySetConfig(out string error)
+
+            if (this.existingConfigSettings == null)
+            {
+                return this.ConfigureWatchmanIntegration(out error);
+            }
+
+            GitProcess.ConfigResult config = null;
+            GitProcess.Result getResult = this.RunGitCommand(
+                process => {
+                    config = process.GetFromLocalConfig("feature.scalar");
+                    return null;
+                },
+                nameof(GitProcess.GetFromLocalConfig)
+            );
+            GitFeatureFlags flags = GitVersion.GetAvailableGitFeatures(this.Context.Tracer);
+            config.TryParseAsString(out string scalar, out error, defaultValue: "true");
+
+            if (scalar.Equals("false"))
+            {
+                GitProcess.Result deleteResult = this.RunGitCommand(
+                    process => process.DeleteFromLocalConfig("core.fsmonitor"),
+                    nameof(GitProcess.DeleteFromLocalConfig)
+                );
+
+                return deleteResult.ExitCodeIsSuccess;
+            }
+            else if (scalar.Equals("experimental")
+                     // Make sure Git supports builtin FS Monitor
+                     && flags.HasFlag(GitFeatureFlags.BuiltinFSMonitor)
+                     // For now, this doesn't work on Linux
+                     && !RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                // ":internal:" is a custom value to specify the builtin
+                // FS Monitor feature.
+                GitProcess.Result setResult = this.RunGitCommand(
+                    process => process.SetInLocalConfig("core.fsmonitor", ":internal:"),
+                    nameof(GitProcess.SetInLocalConfig)
+                );
+
+                return setResult.ExitCodeIsSuccess;
+            }
+            else
+            {
+                return this.ConfigureWatchmanIntegration(out error);
+            }
         }
 
         protected override void PerformMaintenance()
         {
             this.TrySetConfig(out _);
         }
 
-        private bool TrySetConfig(Dictionary<string, string> configSettings, bool isRequired, out string error, bool add = false)
+        private Dictionary<string, GitConfigSetting> GetConfigSettings()
         {
-            Dictionary<string, GitConfigSetting> existingConfigSettings = null;
+            if (this.existingConfigSettings != null)
+            {
+                return this.existingConfigSettings;
+            }
 
             GitProcess.Result getConfigResult = this.RunGitCommand(
-                process => process.TryGetAllConfig(localOnly: isRequired, configSettings: out existingConfigSettings),
+                process => process.TryGetAllConfig(localOnly: true, configSettings: out this.existingConfigSettings),
                 nameof(GitProcess.TryGetAllConfig));
 
+            if (getConfigResult.ExitCodeIsFailure)
+            {
+                this.existingConfigSettings = null;
```

Comment:
If we kept a local variable here for use with the above `out` parameter then we could avoid having to lose the reference to the old instance in the case of failure. Just do a `this.existingConfigSettings = localVar` in the case of success.

Reply:
we also only run this command if `this.existingConfigSettings` is already `null`, so this check at the end means nothing. Will remove.